yellow-paper/docs/public-vm/avm.md

```diff
@@ -77,6 +79,84 @@ MachineState {
+
+The "Accrued Substate", as coined in the [Ethereum Yellow Paper](https://ethereum.github.io/yellowpaper/paper), contains information that is accrued throughout transaction execution to be "acted upon immediately following the transaction." These are append-only arrays containing state that is not relevant to other calls or transactions. Similar to world state, if a message call succeeds, its substate is appended to its calling context, but if it fails its substate is dropped by its caller.
+```
+AccruedSubstate {
+    logs: [],           // append-only
+    l2toL1Messages: [], // append-only
+}
+```
+
+Finally, when a message call halts, it sets the context's "Message Call Results" to communicate results to the caller.
+```
+MessageCallResults {
+    reverted: boolean,
+    output: [] | undefined,
+}
+```
+
+### Context initialization for initial call
+This section outlines AVM context initialization specifically for a **public execution request's initial message call** (_i.e._ not a nested message call). Context initialization for nested message calls will be explained in a later section.
+
+When AVM execution is initiated for a public execution request, the AVM context is initialized as follows:
+```
+context = AVMContext {
+    environment: INITIAL_EXECUTION_ENVIRONMENT,
+    machineState: INITIAL_MACHINE_STATE,
+    accruedSubstate: empty,
+    worldState: <latest world state>,
+    results: INITIAL_MESSAGE_CALL_RESULTS,
+}
+```
+> Note: Since world state persists between transactions, the latest state is injected into a new AVM context.
+
+Given a `PublicCallRequest` and its parent `TxRequest`, these above-listed "`INITIAL_*`" entries are defined as follows:
+```
+INITIAL_EXECUTION_ENVIRONMENT = ExecutionEnvironment {
+    address: PublicCallRequest.contractAddress,
+    storageAddress: PublicCallRequest.CallContext.storageContractAddress,
+    origin: TxRequest.origin,
+    l1GasPrice: TxRequest.l1GasPrice,
+    l2GasPrice: TxRequest.l2GasPrice,
+    calldata: PublicCallRequest.args,
+    sender: PublicCallRequest.CallContext.msgSender,
+    portal: PublicCallRequest.CallContext.portalContractAddress,
+    bytecode: worldState.contracts[PublicCallRequest.contractAddress],
+    blockHeader: <latest block header>,
+    globalVariables: <latest global variable values>
+    messageCallDepth: 0,
+    isStaticCall: PublicCallRequest.CallContext.isStaticCall,
+    isDelegateCall: PublicCallRequest.CallContext.isDelegateCall,
+}
+
+INITIAL_MACHINE_STATE = MachineState {
+    l1GasLeft: TxRequest.l1GasLimit,
+    l2GasLeft: TxRequest.l2GasLimit,
+    pc: 0,
     memory: uninitialized,
 }
+
+INITIAL_MESSAGE_CALL_RESULTS = MessageCallResults {
+    reverted: false,
+    output: undefined,
+}
 ```
+
+> Note: unlike memory in the Ethereum Virtual Machine, uninitialized memory in the AVM is not readable! A memory cell must be written (and therefore [type-tagged](./state-model#types-and-tagged-memory)) before it can be read.
```

Comment:
In the VM circuit I can constrain that memory is initialized with 0 and flag type 0, so that any initial read at memory location (on which we did not write to) returns to zero. If Noir compiler can ensure that we never read from uninitialized memory, we can simply constrain the memory trace that the first memory operation per address is always a write. My point is that we can support both ways. We should just be on the same page with the noir team.

Reply:
Flag type `0` means `uninitialized`, and I thought that any "loads" from an uninitialized cell would fail. Is that true? Could we get rid of the concept of `uninitialized` and just say that all cells start tagged as `field` and initialized to 0?

Reply:
I think enforcing type field initialized with 0 is a bad trade-off. I would consider 2 options:
1) Forbid any uninitialized read
2) Allow read on memory of "uninitialized" words which work with any type and read 0

My point is that the circuit should not have any difficulty to support these variants. That's why I would check with Noir team what is preferable.